content/blog/2023-12-11-bloke/index.md

```diff
@@ -0,0 +1,372 @@
+
+$$
+\text{perf}(x)=\sum\_{i \in \text{instr}(x)} \text { latency }(i)
+$$
+
+where latency $(i)$ is the average latency of an instruction $i$.
+
+## Two-Stage Process
+
+STOKE uses a two-stage process to find optimized programs. STOKE first has the synthesis phase that focus solely on correctness. STOKE then uses the optimization phase to introduce the performance component of the cost function, which allows the search for the fastest program in the correct program space.
+
+In short, the synthesis and the optimization phases have these cost functions:
+
+- synthesis, eq $(x)$; and
+- optimization, eq $(x)+\text{perf}(x)$.
+
+
+## BLOKE
+
+BLOKE is a stochastic superoptimizer for Bril programs. Inspired from the explore-andexploit nature of simulated annealing, we wondered whether gradually introducing the per-
+formance component to STOKE would help us find superoptimizations faster. Therefore, one key difference between BLOKE and STOKE is the introduction of a gradual performance factor into its cost function. This allows us to first explore the space of equivalent programs, then gradually exploit performant programs to superoptimize our intial program.
+
+For simplicity, we only consider loop-free Bril programs with integer return values and only primitive operations. Adding support for more Bril operations such as function calls is future work.
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-04.jpg?height=421&width=578&top_left_y=930&top_left_x=354)
+
+Figure 2: BLOKE optimization of clobber
+
+Figure 2 shows the optimization BLOKE may produce when Figure 1 is the input. We successfully find that the last three lines of Figure 1 are the only necessary lines for an optimized equivalent program.
+
+## Architecture
+
+Figure 3 shows high-level architecture of BLOKE. The user would call the optimizer library which samples using STOKE's random walk method. The sample library can call the Brilirs interpreter to run test cases, or call the bril\_equivalence library to prove that a program in the walk is equivalent to the input program. The equivalence library would lift the Bril programs into $\mathrm{Z} 3$ formulas using the bril2z3 library and call $Z 3$ to prove equivalence or find a counterexample.
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-05.jpg?height=352&width=1482&top_left_y=239&top_left_x=316)
+
+Figure 3: High level BLOKE architecture
+
+## Correctness Cost
+
+Since we have integer outputs, we can use the usual Euclidean metric to define the validation cost similar to STOKE:
+
+$$
+\text{val}(x)=\sum\_{(i, o) \in \tau}|x(i)-o|+\text{err}(x, i)
+$$
+
+Currently we set $\text{err}(x, i)=2$ for all inputs. Varying the error cost based on the error type is future work.
+
+## SMT Bril Equivalence checking
+
+For the verification cost, we adopted Equation 1 and wrote two libraries, bril2z 3 and bril\_equivalence.
```

Comment:
Aha, now the are introduced! Is it possible to move this introduction above, before the first mention of the libraries?

Reply:
Moved the architecture section below the parallelization section